# Data-Driven Decision Making in SQL From DATACAMP
# Notes taken by César Muro

Objectives of data-driven decision-making:
* Information for operational decisions
    * Popularity of actors to decide which movies to invest in
    * Revenue of the last months to estimate budget for short-term investments
* Information for strategic decisions
     * Sucess across countries to decide on market extensions
     * Longterm development of revenue for long term investments
   
Key perfomance indicators hel a company define and monitor success

## Filter and ordering  
A main part of querying data is to filter those records which are relevant for a certain research question. For future applications, it is also essential to bring tables in a particular order.  

Operators in where clause include comparison operators, between, is null, is not null.  

In [1]:
--  Select all movies rented on October 9th, 2018.
SELECT *
FROM movies.renting
where date_renting ='2018-10-09'; -- Movies rented on October 9th, 2018

renting_id  customer_id  movie_id  rating              date_renting
0           1           41         8     NaN 2018-10-09 00:00:00+00:00
1           6           50        71     7.0 2018-10-09 00:00:00+00:00

In [1]:
-- Add your own queries here
SELECT *
FROM movies.actors
LIMIT 5

actor_id                name  year_of_birth nationality  gender
0         1       Abbie Cornish           1982   Australia  female
1         2        Adam Sandler           1966         USA    male
2         3           Al Pacino           1940         USA    male
3         4           Amy Adams           1974         USA  female
4         5  Andrea Riseborough           1981     British  female

In [2]:
-- Select all records of movie rentals between beginning of April 2018 till end of August 2018.
SELECT *
FROM movies.renting
WHERE date_renting between '2018-04-01' and '2018-08-31'; -- from beginning April 2018 to end August 2018

renting_id  customer_id  movie_id  rating              date_renting
0             3          108        45     4.0 2018-06-08 00:00:00+00:00
1             8           73        65    10.0 2018-06-05 00:00:00+00:00
2            12           52        65    10.0 2018-06-29 00:00:00+00:00
3            14            8        29     NaN 2018-08-03 00:00:00+00:00
4            16           59        59     NaN 2018-08-12 00:00:00+00:00
..          ...          ...       ...     ...                       ...
133         561          113        36     5.0 2018-06-29 00:00:00+00:00
134         563           29         5     NaN 2018-05-11 00:00:00+00:00
135         566           28        45     7.0 2018-05-29 00:00:00+00:00
136         569           78        26     NaN 2018-08-16 00:00:00+00:00
137         575           45        31    10.0 2018-06-09 00:00:00+00:00

[138 rows x 5 columns]

In [3]:
--Put the most recent records of movie rentals on top of the resulting table and order them in decreasing order.
SELECT *
FROM movies.renting
WHERE date_renting BETWEEN '2018-04-01' AND '2018-08-31'
order by date_renting desc; -- Order by recency in decreasing order

renting_id  customer_id  movie_id  rating              date_renting
0            76          112        10     NaN 2018-08-31 00:00:00+00:00
1           279           21        62     NaN 2018-08-30 00:00:00+00:00
2           551           86        13     8.0 2018-08-29 00:00:00+00:00
3           454          118        32     6.0 2018-08-28 00:00:00+00:00
4           463           29        68     NaN 2018-08-28 00:00:00+00:00
..          ...          ...       ...     ...                       ...
133          57           25        29     6.0 2018-04-05 00:00:00+00:00
134          28          113        40     7.0 2018-04-05 00:00:00+00:00
135         303          121        58     NaN 2018-04-04 00:00:00+00:00
136         207           52        25     5.0 2018-04-03 00:00:00+00:00
137         218           94        34     6.0 2018-04-03 00:00:00+00:00

[138 rows x 5 columns]

In [4]:
--    Select all movies which are not dramas.
SELECT *
FROM movies.movies
Where genre <> 'Drama'; -- All genres except drama

movie_id  ... renting_price
0          1  ...          2.09
1          3  ...          2.59
2          5  ...          2.59
3          6  ...          2.69
4          7  ...          1.89
5          8  ...          2.59
6         11  ...          1.79
7         12  ...          2.39
8         13  ...          1.79
9         14  ...          1.69
10        15  ...          1.69
11        17  ...          1.99
12        20  ...          2.29
13        21  ...          1.99
14        22  ...          2.59
15        24  ...          2.89
16        28  ...          1.69
17        31  ...          2.69
18        35  ...          1.79
19        43  ...          2.09
20        44  ...          2.49
21        47  ...          2.69
22        48  ...          2.89
23        49  ...          1.59
24        51  ...          2.99
25        53  ...          2.49
26        54  ...          2.79
27        56  ...          2.09
28        59  ...          2.39
29        61  ...          1.99
30        63  ...          2.69
31        65  ...          1.79
32        68  ...          2.39
33        69  ...          2.49
34        71  ...          1.79

[35 rows x 6 columns]

In [5]:
-- Select the movies 'Showtime', 'Love Actually' and 'The Fighter'
SELECT *
FROM movies.movies
WHERE title in ('Showtime','Love Actually','The Fighter'); -- Select all movies with the given titles

movie_id          title   genre  runtime  year_of_release  renting_price
0        11       Showtime  Comedy       95             2002           1.79
1        20  Love Actually  Comedy      135             2003           2.29
2        53    The Fighter   Other      116             2010           2.49

In [6]:
SELECT *
FROM movies.movies
order by renting_price ; -- Order the movies by increasing renting price

movie_id  ... renting_price
0         42  ...          1.49
1         16  ...          1.59
2         49  ...          1.59
3         37  ...          1.59
4         36  ...          1.59
..       ...  ...           ...
66        32  ...          2.89
67        52  ...          2.89
68        25  ...          2.89
69         9  ...          2.99
70        51  ...          2.99

[71 rows x 6 columns]

In [7]:
--Select from table renting all movie rentals from 2018.
--Filter only those records which have a movie rating

SELECT *
FROM movies.renting
WHERE date_renting between '2018-01-01' and '2018-12-31' -- Renting in 2018
AND rating is not null; -- Rating exists

renting_id  customer_id  movie_id  rating              date_renting
0             3          108        45       4 2018-06-08 00:00:00+00:00
1             4           39        66       8 2018-10-22 00:00:00+00:00
2             6           50        71       7 2018-10-09 00:00:00+00:00
3             8           73        65      10 2018-06-05 00:00:00+00:00
4            12           52        65      10 2018-06-29 00:00:00+00:00
..          ...          ...       ...     ...                       ...
158         561          113        36       5 2018-06-29 00:00:00+00:00
159         565           63        53      10 2018-10-25 00:00:00+00:00
160         566           28        45       7 2018-05-29 00:00:00+00:00
161         571           92        28       9 2018-11-22 00:00:00+00:00
162         575           45        31      10 2018-06-09 00:00:00+00:00

[163 rows x 5 columns]

## Aggregations summarizing data  
For decision making, it is usually not important to look at individual records, but rather to examine summaries of certain groups.  
With the aggregate functions the null values are ignores. 

In [8]:
--    Count the number of customers born in the 80s.

SELECT count(*)-- Count the total number of customers
FROM movies.customers
WHERE date_of_birth between '1980-01-01' and '1980-12-31'; -- Select customers born between 1980-01-01 and 1989-12-31

count
0      4

In [9]:
SELECT count(*)   -- Count the total number of customers
FROM movies.customers
WHERE country='Germany'; -- Select all customers from Germany

count
0      0

In [10]:
--    Count the number of countries where MovieNow has customers.
SELECT count(DISTINCT country)  -- Count the number of countries
FROM movies.customers

count
0     11

In [11]:
SELECT min(rating) as min_rating, -- Calculate the minimum rating and use alias min_rating
	   max(rating) max_rating, -- Calculate the maximum rating and use alias max_rating
	   avg(rating) as avg_rating, -- Calculate the average rating and use alias avg_rating
	count(rating) number_ratings -- Count the number of ratings and use alias number_ratings
FROM movies.renting
WHERE movie_id=25; -- Select all records of the movie with ID 25

min_rating  max_rating  avg_rating  number_ratings
0           5          10         7.5               8

Examining annual rentals

You are asked to provide a report about the development of the company. Specifically, your manager is interested in the total number of movie rentals, the total number of ratings and the average rating of all movies since the beginning of 2019.

In [12]:
-- Select all records of movie rentals since January 1st 2019
Select *
from movies.renting
WHERE date_renting between '2019-01-01' and now();

renting_id  customer_id  movie_id  rating              date_renting
0             5          104        15     7.0 2019-03-18 00:00:00+00:00
1            17           22        46    10.0 2019-02-16 00:00:00+00:00
2            18           36        39    10.0 2019-03-20 00:00:00+00:00
3            27            7        36     NaN 2019-03-14 00:00:00+00:00
4            32            8        42    10.0 2019-02-13 00:00:00+00:00
..          ...          ...       ...     ...                       ...
154         560           11        49     6.0 2019-04-18 00:00:00+00:00
155         562           33        18     NaN 2019-03-10 00:00:00+00:00
156         567           28        49     8.0 2019-02-24 00:00:00+00:00
157         572            4        61     6.0 2019-04-16 00:00:00+00:00
158         574            7        14     NaN 2019-02-08 00:00:00+00:00

[159 rows x 5 columns]

In [14]:
SELECT 
	COUNT(*) AS number_renting,
	AVG(rating) AS average_rating, 
    COUNT(rating) AS number_ratings -- Add the total number of ratings here.
FROM movies.renting
WHERE date_renting >= '2019-01-01';

number_renting  average_rating  number_ratings
0             159        7.946237              93

## Grouping movies
Using the group by clause allows us to apply aggregations to groups in a table

In [1]:
--    Create a table with a row for each country and columns for the country name and the date when the first customer account was created.
 --   Use the alias first_account for the column with the dates.
--    Order by date in ascending order.


SELECT country, min(date_account_start) as first_account
FROM movies.customers
GROUP BY country
ORDER BY first_account;

country             first_account
0         France 2017-01-13 00:00:00+00:00
1        Hungary 2017-01-18 00:00:00+00:00
2        Belgium 2017-01-28 00:00:00+00:00
3       Slovenia 2017-01-31 00:00:00+00:00
4          Spain 2017-02-14 00:00:00+00:00
5          Italy 2017-02-28 00:00:00+00:00
6         Poland 2017-03-03 00:00:00+00:00
7   Great Britan 2017-03-31 00:00:00+00:00
8        Denmark 2017-04-30 00:00:00+00:00
9            USA 2017-09-13 00:00:00+00:00
10       Austria 2017-11-22 00:00:00+00:00

In [2]:
SELECT movie_id, -- Group the data in the table renting by movie_id and report the ID and the average rating.
       avg(rating)    -- Calculate average rating per movie
FROM movies.renting
group by movie_id;

movie_id       avg
0         54  8.166667
1         29  8.000000
2         71  8.000000
3         68  6.333333
4          4  7.888889
..       ...       ...
66        27  8.285714
67        23  8.000000
68        56  5.666667
69        58  8.333333
70         8  8.000000

[71 rows x 2 columns]

In [3]:
SELECT movie_id, 
       AVG(rating) AS  avg_rating, -- Use as alias avg_rating
       count(rating) AS number_rating,                -- Add column for number of ratings with alias number_rating
       count(movie_id) as number_renting                 -- Add column for number of movie rentals with alias number_renting
FROM movies.renting
GROUP BY movie_id;

movie_id  avg_rating  number_rating  number_renting
0         54    8.166667              6               9
1         29    8.000000              7              11
2         71    8.000000              4               6
3         68    6.333333              3               7
4          4    7.888889              9              14
..       ...         ...            ...             ...
66        27    8.285714              7              13
67        23    8.000000              7              11
68        56    5.666667              3               4
69        58    8.333333              3              10
70         8    8.000000              1               5

[71 rows x 4 columns]

In [4]:
--     Group the data in the table renting by customer_id and report the customer_id, the average rating, the number of ratings and the number of movie rentals.
--    Select only customers with more than 7 movie rentals.
--    Order the resulting table by the average rating in ascending order.

SELECT customer_id, -- Report the customer_id
      avg(rating),  -- Report the average rating per customer
      count(rating),  -- Report the number of ratings per customer
      count(*) -- Report the number of movie rentals per customer
FROM movies.renting
GROUP BY customer_id
HAVING count(*) > 7 -- Select only customers with more than 7 movie rentals
ORDER BY avg(rating); -- Order by the average rating in ascending order

customer_id       avg count count
0           104  6.250000     8     8
1            28  6.714286    11    11
2           111  7.000000    10    10
3           113  7.000000    15    15
4            25  7.200000    10    10
5            21  7.333333    14    14
6            92  7.571429    11    11
7            49  7.625000    13    13
8            35  7.666667     9     9
9            52  7.875000     9     9
10          108  8.000000    10    10
11          114  8.000000    11    11
12           66  8.000000    10    10
13           93  8.400000     8     8
14           26  8.400000    10    10
15           94  8.714286    10    10
16          121  8.750000     9     9
17            7  9.000000     9     9
18           61  9.200000     8     8
19            8  9.250000     8     8
20           99  9.250000     8     8
21           65  9.333333     9     9

## Joining movie ratings with customer data  
Join queries allow us to combine or join data from multiple tables.  
Left join is an outer table; keep all rows from the left, mathc with rows in the right table.

In [5]:
--   Augment the table renting with all columns from the table customers with a LEFT JOIN.
--   Use as alias' for the tables r and c respectively.

SELECT * -- Join renting with customers
FROM movies.renting as r
LEFT JOIN movies.customers as c
ON r.customer_id=c.customer_id;

renting_id  ...        date_account_start
0             1  ... 2017-06-12 00:00:00+00:00
1             2  ... 2017-01-28 00:00:00+00:00
2             3  ... 2017-03-13 00:00:00+00:00
3             4  ... 2018-01-19 00:00:00+00:00
4             5  ... 2017-04-12 00:00:00+00:00
..          ...  ...                       ...
573         574  ... 2018-05-12 00:00:00+00:00
574         575  ... 2017-08-09 00:00:00+00:00
575         576  ... 2017-11-06 00:00:00+00:00
576         577  ... 2017-05-22 00:00:00+00:00
577         578  ... 2017-05-22 00:00:00+00:00

[578 rows x 11 columns]

In [6]:
SELECT avg(r.rating) -- Average ratings of customers from Belgium
FROM movies.renting AS r
LEFT JOIN movies.customers AS c
ON r.customer_id = c.customer_id
WHERE c.country='Belgium';

avg
0  8.9

In [7]:
--    First, you need to join movies on renting to include the renting_price from the movies table for each renting record.
 --   Use as alias' for the tables m and r respectively.

SELECT *
FROM movies.renting AS r
LEFT join movies.movies AS m -- Choose the correct join statment
ON r.movie_id=m.movie_id;

renting_id  customer_id  movie_id  ...  runtime year_of_release  renting_price
0             1           41         8  ...       91            2002           2.59
1             2           10        29  ...      122            2005           2.79
2             3          108        45  ...       96            2008           2.39
3             4           39        66  ...      142            2012           1.59
4             5          104        15  ...      128            2003           1.69
..          ...          ...       ...  ...      ...             ...            ...
573         574            7        14  ...      115            2003           1.69
574         575           45        31  ...      157            2005           2.69
575         576           38        25  ...      108            2004           2.89
576         577           35        65  ...       96            2012           1.79
577         578           35        16  ...      135            2003           1.59

[578 rows x 11 columns]

In [8]:
SELECT 
	sum(m.renting_price), -- Get the revenue from movie rentals
	count(*), -- Count the number of rentals
	count(distinct r.customer_id) -- Count the number of customers
FROM movies.renting AS r
LEFT JOIN movies.movies AS m
ON r.movie_id = m.movie_id;

sum count count
0  1275.72   116   116

In [9]:
SELECT 
	SUM(m.renting_price), 
	COUNT(*), 
	COUNT(DISTINCT r.customer_id)
FROM movies.renting AS r
LEFT JOIN movies.movies AS m
ON r.movie_id = m.movie_id
-- Only look at movie rentals in 2018
WHERE date_renting BETWEEN '2018-01-01' AND '2018-12-31' ;

sum count count
0  658.02    93    93

In [10]:
SELECT a.name, -- Create a list of movie titles and actor names Make sure that each combination of actor and movie appears only once.
     m.title  
FROM movies.actsin as ai
LEFT JOIN movies.movies AS m
ON m.movie_id = ai.movie_id
LEFT JOIN movies.actors AS a
ON a.actor_id = ai.actor_id;

name                             title
0       Abbie Cornish                             Candy
1        Adam Sandler                     Jack and Jill
2           Al Pacino                            Simone
3           Al Pacino                       The Recruit
4           Al Pacino                 Two for the Money
..                ...                               ...
208       Viola Davis                          The Help
209  Violante Placido  Ghost Rider: Spirit of Vengeance
210         will.i.am                             Rio 2
211   William H. Macy                      The Sessions
212   Zooey Deschanel                 Failure to Launch

[213 rows x 2 columns]

## Money spent per customer with sub-queries  
Learn how to calculate how much money each customer spent on MovieNow rentals by using sub-queries. 

In [11]:
SELECT m.title, -- Use a join to get the movie title and price for each movie rental
       m.renting_price
FROM movies.renting AS r
LEFT JOIN movies.movies AS m
ON m.movie_id=r.movie_id;

title  renting_price
0                      Waking Up in Reno           2.59
1                      Two for the Money           2.79
2                     Burn After Reading           2.39
3                       The Hunger Games           1.59
4                 Something's Gotta Give           1.69
..                                   ...            ...
573                          The Recruit           1.69
574  Harry Potter and the Goblet of Fire           2.69
575   Bridget Jones - The Edge of Reason           2.89
576     Ghost Rider: Spirit of Vengeance           1.79
577                            25th Hour           1.59

[578 rows x 2 columns]

In [12]:
SELECT rm.title, -- Report the income from movie rentals for each movie 
       SUM(rm.renting_price) AS income_movie
FROM
       (SELECT m.title, 
               m.renting_price
       FROM movies.renting AS r
       LEFT JOIN movies.movies AS m
       ON r.movie_id=m.movie_id) AS rm
GROUP BY rm.title
ORDER BY income_movie DESC; -- Order the result by decreasing income

title  income_movie
0        Bridget Jones - The Edge of Reason         37.57
1                                 Fair Game         34.68
2                               The Kingdom         31.35
3                         Two for the Money         30.69
4                          Django Unchained         29.59
..                                      ...           ...
66  Harry Potter and the Chamber of Secrets          8.95
67                            Jack and Jill          8.36
68                             The Invasion          8.36
69                          The Human Stain          7.96
70                          The Other Woman          3.38

[71 rows x 2 columns]

In [13]:
--    Create a subsequent SELECT statements in the FROM clause to get all information about actors from the USA.
--    Give the subsequent SELECT statement the alias a.
--    Report for actors from the USA the year of birth of the oldest and the year of birth of the youngest actor and actress.


SELECT  a.gender,-- Report for male and female actors from the USA 
       min(a.year_of_birth), -- The year of birth of the oldest actor
       max(a.year_of_birth) -- The year of birth of the youngest actor
FROM
   (SELECT * -- Use a subsequen SELECT to get all information about actors from the USA
   from movies.actors
   where nationality='USA') a -- Give the table the name a
   group by a.gender;

gender   min   max
0  female  1945  1993
1    male  1930  1992

## Identify favorite actors of customer groups
Complex queries to answer the question: Identify favorite movies for a group of customers
Which is the favorite movie on MovieNow? Answer this question for a specific group of customers: for all customers born in the 70s

In [14]:
SELECT * --Augment the table renting with customer information and information about the movies.
FROM movies.renting AS r
LEFT JOIN movies.customers as c   -- Add customer information
on c.customer_id=r.customer_id
LEFT JOIN movies.movies as m   -- Add movie information
on m.movie_id=r.movie_id;

renting_id  customer_id  movie_id  ...  runtime year_of_release  renting_price
0             1         41.0         8  ...       91            2002           2.59
1             2         10.0        29  ...      122            2005           2.79
2             3        108.0        45  ...       96            2008           2.39
3             4         39.0        66  ...      142            2012           1.59
4             5        104.0        15  ...      128            2003           1.69
..          ...          ...       ...  ...      ...             ...            ...
573         574          7.0        14  ...      115            2003           1.69
574         575         45.0        31  ...      157            2005           2.69
575         576         38.0        25  ...      108            2004           2.89
576         577         35.0        65  ...       96            2012           1.79
577         578         35.0        16  ...      135            2003           1.59

[578 rows x 17 columns]

In [15]:
SELECT *
FROM movies.renting AS r
LEFT JOIN movies.customers AS c
ON c.customer_id = r.customer_id
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
WHERE date_of_birth between '1970-01-01' and '1979-12-31'; -- Select customers born in the 70s

renting_id  customer_id  movie_id  ...  runtime year_of_release  renting_price
0             4           39        66  ...      142            2012           1.59
1            11           61        61  ...      130            2011           1.99
2            15           13         5  ...      178            2001           2.59
3            17           22        46  ...      101            2008           1.79
4            18           36        39  ...      135            2007           1.79
..          ...          ...       ...  ...      ...             ...            ...
160         547           65        63  ...      165            2012           2.69
161         560           11        49  ...      153            2009           1.59
162         561          113        36  ...      129            2006           1.59
163         565           63        53  ...      116            2010           2.49
164         576           38        25  ...      108            2004           2.89

[165 rows x 17 columns]

In [17]:
SELECT m.title, 
COUNT(*),
AVG(r.rating)
FROM movies.renting AS r
LEFT JOIN movies.customers AS c
ON c.customer_id = r.customer_id
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
WHERE c.date_of_birth BETWEEN '1970-01-01' AND '1979-12-31'
GROUP BY m.title
HAVING  count(*) >1 -- Remove movies with only one rental; -- Order with highest rating first
order by avg(r.rating) desc;

title  count        avg
0                                        Showtime      5        NaN
1   Harry Potter and the Deathly Hallows – Part 2      2        NaN
2                               Waking Up in Reno      2        NaN
3                                             Ray      2        NaN
4                                Django Unchained      4  10.000000
5                                     The Fighter      4  10.000000
6                           One Night at McCool's      2  10.000000
7                          No Country for Old Men      3  10.000000
8                      The Fellowship of the Ring      2  10.000000
9                                   I'm Not There      2  10.000000
10                               Secondhand Lions      4   9.666667
11                                  Winter's Bone      2   9.500000
12                                    Fool's Gold      4   9.333333
13                                      25th Hour      2   9.000000
14                             Burn After Reading      2   9.000000
15                                  Morning Glory      2   9.000000
16                                       The Help      3   9.000000
17                              Failure to Launch      4   8.666667
18                                    Young Adult      4   8.500000
19                                   Sleepwalking      2   8.500000
20                                    Black Death      3   8.500000
21             Bridget Jones - The Edge of Reason      6   8.250000
22                           Welcome to Mooseport      4   8.000000
23         Harry Potter and the Half-Blood Prince      3   8.000000
24                           The Royal Tenenbaums      2   8.000000
25                                      Fair Game      3   8.000000
26                                        Monster      5   8.000000
27                              The Frozen Ground      4   7.500000
28       Harry Potter and the Prisoner of Azkaban      4   7.500000
29      Harry Potter and the Order of the Phoenix      3   7.500000
30                              Two for the Money      4   7.333333
31                                   A Good Woman      4   7.333333
32                                   Training Day      3   7.000000
33                                          Candy      3   7.000000
34                                         Capote      3   7.000000
35                                 Shall We Dance      4   7.000000
36                                     Happy Feet      3   7.000000
37                                 V for Vendetta      2   7.000000
38                                 The Two Towers      3   7.000000
39                                    The Kingdom      5   6.666667
40                               The Hunger Games      2   6.500000
41                             World Trade Center      3   6.500000
42                                         Simone      5   6.000000
43                              Then She Found Me      4   6.000000
44  Harry Potter and the Deathly Hallows – Part 1      2   6.000000
45                         Something's Gotta Give      3   6.000000
46                                    The Recruit      2   5.000000

Identify favorite actors for Spain

You're now going to explore actor popularity in Spain. Use as alias the first letter of the table, except for the table actsin use ai instead.

In [18]:
SELECT *
FROM movies.renting as r 
LEFT JOIN movies.customers as c  -- Augment table renting with information about customers 
on r.customer_id=c.customer_id
LEFT JOIN  movies.actsin as ai -- Augment the table renting with the table actsin
on r.movie_id=ai.movie_id
LEFT JOIN movies.actors as a  -- Augment table renting with information about actors
on ai.actor_id=a.actor_id;

renting_id  customer_id  ...       nationality  gender
0              1         41.0  ...               USA    male
1              1         41.0  ...           British  female
2              1         41.0  ...      South Africa  female
3              2         10.0  ...               USA  female
4              2         10.0  ...               USA    male
...          ...          ...  ...               ...     ...
1729         577         35.0  ...               USA    male
1730         577         35.0  ...  Northern Ireland    male
1731         578         35.0  ...               USA    male
1732         578         35.0  ...               USA    male
1733         578         35.0  ...            Canada    male

[1734 rows x 19 columns]

In [19]:
SELECT a.name,  c.gender,
       COUNT(*) AS number_views, 
       AVG(r.rating) AS avg_rating
FROM movies.renting as r
LEFT JOIN movies.customers AS c
ON r.customer_id = c.customer_id
LEFT JOIN movies.actsin as ai
ON r.movie_id = ai.movie_id
LEFT JOIN movies.actors as a
ON ai.actor_id = a.actor_id
WHERE c.country='Spain'-- Select only customers from Spain
GROUP BY a.name, c.gender
HAVING AVG(r.rating) IS NOT NULL 
  AND COUNT(*) > 5 
ORDER BY avg_rating DESC, number_views DESC;

name  gender  number_views  avg_rating
0  Catherine Keener  female             6         8.0
1       Emma Watson    male             7         7.6
2  Daniel Radcliffe    male             7         7.6
3      Rupert Grint    male             7         7.6

Catherine Keener is the favorite actress among female Spain customers and that male customers from Spain like the actors from Harry Potter best: Emma Watson, Daniel Radcliffe and Rupert Grint.

 Calculate the number of movie rentals.  
 Calculate the average rating.  
 Calculate the revenue from movie rentals.  
 Report these KPIs for each country.  


In [1]:
SELECT 
	c.country,                   -- For each country report
	count(*) AS number_renting, -- The number of movie rentals
	avg(r.rating) AS average_rating, -- The average rating
	sum(m.renting_price) AS revenue         -- The revenue from movie rentals
FROM movies.renting AS r
LEFT JOIN movies.customers AS c
ON c.customer_id = r.customer_id
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
WHERE date_renting >= '2019-01-01'
group by c.country;

country  number_renting  average_rating  revenue
0           None               1       10.000000     1.79
1          Spain              26        8.076923    57.94
2   Great Britan               9        7.200000    17.91
3        Austria               4        6.000000     8.16
4          Italy              30        8.235294    69.90
5         Poland              21        8.125000    49.29
6       Slovenia              14        8.000000    30.26
7        Hungary              11        7.166667    25.89
8        Denmark               7        7.714286    14.63
9        Belgium               7        9.000000    13.13
10           USA               6        9.000000    14.54
11        France              23        7.533333    51.07

## Nested query  
It is a special kind of subquery. A nested query is a query where a complete select clause appear in the where clause or the having clause of another query. Inner query returns single or multiple values.

Your manager wants you to make a list of movies excluding those which are hardly ever watched. This list of movies will be used for advertising. List all movies with more than 5 views using a nested query which is a powerful tool to implement selection conditions.

In [2]:
SELECT movie_id -- Select all movies IDs with more than 5 views
FROM movies.renting
GROUP BY movie_id
HAVING COUNT(*) > 5

movie_id
0         54
1         29
2         71
3         68
4          4
5         34
6         51
7         70
8         52
9         67
10        63
11        10
12        35
13        45
14         6
15        39
16        69
17        36
18        31
19        50
20        14
21        66
22        22
23         2
24        16
25        62
26        11
27        44
28        42
29        41
30        40
31        46
32        53
33        32
34         7
35        38
36        15
37        26
38        12
39        24
40        57
41        19
42        61
43        25
44        30
45        49
46        47
47         3
48        28
49        37
50        33
51         1
52         5
53        18
54        64
55        27
56        23
57        58

In [3]:
SELECT * --  Select all information about movies with more than 5 views.
FROM movies.movies
WHERE movie_id in -- Select movie IDs from the inner query
	(SELECT movie_id
	FROM movies.renting
	GROUP BY movie_id
	HAVING COUNT(*) > 5)

movie_id  ... renting_price
0          1  ...          2.09
1          2  ...          2.19
2          3  ...          2.59
3          4  ...          1.79
4          5  ...          2.59
5          6  ...          2.69
6          7  ...          1.89
7         10  ...          2.69
8         11  ...          1.79
9         12  ...          2.39
10        14  ...          1.69
11        15  ...          1.69
12        16  ...          1.59
13        18  ...          1.69
14        19  ...          2.09
15        22  ...          2.59
16        23  ...          1.69
17        24  ...          2.89
18        25  ...          2.89
19        26  ...          2.59
20        27  ...          2.09
21        28  ...          1.69
22        29  ...          2.79
23        30  ...          2.39
24        31  ...          2.69
25        32  ...          2.89
26        33  ...          1.59
27        34  ...          1.99
28        35  ...          1.79
29        36  ...          1.59
30        37  ...          1.59
31        38  ...          2.49
32        39  ...          1.79
33        40  ...          1.99
34        41  ...          2.09
35        42  ...          1.49
36        44  ...          2.49
37        45  ...          2.39
38        46  ...          1.79
39        47  ...          2.69
40        49  ...          1.59
41        50  ...          2.69
42        51  ...          2.99
43        52  ...          2.89
44        53  ...          2.49
45        54  ...          2.79
46        57  ...          1.99
47        58  ...          2.49
48        61  ...          1.99
49        62  ...          2.29
50        63  ...          2.69
51        64  ...          2.89
52        66  ...          1.59
53        67  ...          1.59
54        68  ...          2.39
55        69  ...          2.49
56        70  ...          1.59
57        71  ...          1.79

[58 rows x 6 columns]

In [4]:
SELECT *
FROM movies.customers
where customer_id in    -- Select all customers with more than 10 movie rentals
	(SELECT customer_id
	FROM movies.renting
	GROUP BY customer_id
	HAVING count(*) > 10);

customer_id  ...        date_account_start
0           21  ... 2017-06-11 00:00:00+00:00
1           28  ... 2017-02-04 00:00:00+00:00
2           49  ... 2017-01-18 00:00:00+00:00
3           92  ... 2017-09-21 00:00:00+00:00
4          113  ... 2017-06-13 00:00:00+00:00
5          114  ... 2017-02-14 00:00:00+00:00

[6 rows x 6 columns]

In [5]:
SELECT movie_id, -- Select movie IDs and calculate the average rating 
       avg(rating)
FROM movies.renting
group by movie_id
having avg(rating) >          -- Of movies with rating above average
	(SELECT AVG(rating)
	FROM movies.renting);

movie_id        avg
0         54   8.166667
1         29   8.000000
2         71   8.000000
3         34   8.000000
4         51   8.428571
5         63   9.333333
6         35   8.000000
7          6   8.800000
8         39   8.000000
9         50   8.000000
10        60   8.500000
11        22   8.666667
12        13   8.000000
13        65   9.000000
14        16   8.600000
15        62   8.250000
16        42   9.600000
17        40   8.166667
18        53   8.000000
19         9   8.333333
20         7   8.000000
21        38   8.666667
22        48  10.000000
23        26   8.333333
24        24   8.400000
25        57   8.833333
26        61   8.000000
27        49   8.636364
28        47   8.750000
29         3   9.000000
30        17   8.000000
31        28   8.000000
32        33   8.400000
33         5   9.750000
34        18   9.000000
35        64   8.250000
36        27   8.285714
37        23   8.000000
38        58   8.333333
39         8   8.000000

In [6]:
SELECT title -- Report the movie titles of all movies with average rating higher than the total average
FROM movies.movies
WHERE movie_id in
	(SELECT movie_id
	 FROM movies.renting
     GROUP BY movie_id
     HAVING AVG(rating) > 
		(SELECT AVG(rating)
		 FROM movies.renting));

title
0                                 What Women Want
1                      The Fellowship of the Ring
2        Harry Potter and the Philosopher's Stone
3                            The Royal Tenenbaums
4                               Waking Up in Reno
5                          11'09''01 September 11
6         Harry Potter and the Chamber of Secrets
7                                       25th Hour
8                                 The Human Stain
9                                Secondhand Lions
10                            Imagining Argentina
11                                 Shall We Dance
12                           Welcome to Mooseport
13                                            Ray
14                                        Monster
15       Harry Potter and the Prisoner of Azkaban
16                              Two for the Money
17                                 V for Vendetta
18                              Failure to Launch
19                                     Happy Feet
20                                 The Lake House
21                                  I'm Not There
22                              Then She Found Me
23                         No Country for Old Men
24                                    Fool's Gold
25                                      Astro Boy
26         Harry Potter and the Half-Blood Prince
27                                  Winter's Bone
28                                  Morning Glory
29                                    The Fighter
30  Harry Potter and the Deathly Hallows – Part 1
31                                    Young Adult
32                                       The Help
33                                    Black Death
34  Harry Potter and the Deathly Hallows – Part 2
35                                   The Sessions
36                               Django Unchained
37                                    Cloud Atlas
38               Ghost Rider: Spirit of Vengeance
39                                          Rio 2

## Correlated nested queries  
In a correlated neste query the condition in the where clause of the inner query references some column of a table declared in the outer query.  The nested query is then evaluated once for each row in the outer query.  
* condition in the where clause of the inner query
* references some column of a table in the outer query  


In [7]:
-- Count movie rentals of customer 45
-- First, count number of movie rentals for customer with customer_id=45. Give the table renting the alias r
select count(*)
from movies.renting as r
where customer_id=45;

count
0      5

In [8]:
-- Select customers with less than 5 movie rentals
-- Now select all columns from the customer table where the number of movie rentals is smaller than 5.
SELECT *
FROM movies.customers as c
WHERE 5 >
	(SELECT count(*)
	FROM movies.renting as r
	WHERE r.customer_id = c.customer_id);

customer_id  ...        date_account_start
0             2  ... 2018-10-15 00:00:00+00:00
1             3  ... 2019-02-14 00:00:00+00:00
2             4  ... 2017-11-22 00:00:00+00:00
3             5  ... 2018-12-16 00:00:00+00:00
4             6  ... 2018-08-26 00:00:00+00:00
..          ...  ...                       ...
61          117  ... 2018-11-09 00:00:00+00:00
62          119  ... 2019-03-06 00:00:00+00:00
63          120  ... 2018-08-06 00:00:00+00:00
64          122  ... 2018-11-27 00:00:00+00:00
65          123  ... 2018-12-30 00:00:00+00:00

[66 rows x 6 columns]

In [9]:
SELECT *
FROM movies.customers as c
WHERE 4 > -- Select all customers with a minimum rating smaller than 4 
	(SELECT MIN(rating)
	FROM movies.renting AS r
	WHERE r.customer_id = c.customer_id);

customer_id  ...        date_account_start
0           28  ... 2017-02-04 00:00:00+00:00
1           41  ... 2017-06-12 00:00:00+00:00
2           86  ... 2017-12-15 00:00:00+00:00
3          120  ... 2018-08-06 00:00:00+00:00

[4 rows x 6 columns]

In [1]:
SELECT *
FROM movies.movies as m
WHERE 5 < -- Select all movies with more than 5 ratings
	(SELECT count(r.rating)
	FROM movies.renting as r
	WHERE r.movie_id=m.movie_id);

movie_id  ... renting_price
0          4  ...          1.79
1         10  ...          2.69
2         12  ...          2.39
3         14  ...          1.69
4         22  ...          2.59
5         23  ...          1.69
6         25  ...          2.89
7         27  ...          2.09
8         28  ...          1.69
9         29  ...          2.79
10        34  ...          1.99
11        36  ...          1.59
12        40  ...          1.99
13        41  ...          2.09
14        44  ...          2.49
15        45  ...          2.39
16        49  ...          1.59
17        50  ...          2.69
18        51  ...          2.99
19        52  ...          2.89
20        54  ...          2.79
21        57  ...          1.99
22        61  ...          1.99
23        63  ...          2.69

[24 rows x 6 columns]

In [2]:
SELECT *
FROM movies.movies AS m
WHERE 8 < -- Select all movies with an average rating higher than 8
	(SELECT avg(r.rating)
	FROM movies.renting AS r
	WHERE r.movie_id = m.movie_id);

movie_id  ... renting_price
0          3  ...          2.59
1          5  ...          2.59
2          6  ...          2.69
3          9  ...          2.99
4         16  ...          1.59
5         18  ...          1.69
6         22  ...          2.59
7         24  ...          2.89
8         26  ...          2.59
9         27  ...          2.09
10        33  ...          1.59
11        38  ...          2.49
12        40  ...          1.99
13        42  ...          1.49
14        47  ...          2.69
15        48  ...          2.89
16        49  ...          1.59
17        51  ...          2.99
18        54  ...          2.79
19        57  ...          1.99
20        58  ...          2.49
21        60  ...          2.69
22        62  ...          2.29
23        63  ...          2.69
24        64  ...          2.89
25        65  ...          1.79

[26 rows x 6 columns]

## Queries with exists
The EXISTS function is another handy feature in SQL.  
* It is a special case of a correlated nested query.  
* Used to check if result of a correlated nested query is empty.  
* It returns: TRUE/FALSE
* Column specified in select component not considered - use select *  

The function not exists  
* TRUE: table is empty  -> row of the outer query is selected 

Customers with at least one rating

Having active customers is a key performance indicator for MovieNow. Make a list of customers who gave at least one rating.

In [3]:
SELECT *
FROM movies.customers as c -- Select all customers with at least one rating
WHERE EXISTS
	(SELECT *
	FROM movies.renting AS r
	WHERE rating IS NOT NULL 
	AND r.customer_id = c.customer_id );

customer_id  ...        date_account_start
0              2  ... 2018-10-15 00:00:00+00:00
1              4  ... 2017-11-22 00:00:00+00:00
2              7  ... 2018-05-12 00:00:00+00:00
3              8  ... 2018-02-08 00:00:00+00:00
4             10  ... 2017-01-28 00:00:00+00:00
..           ...  ...                       ...
98           118  ... 2017-11-24 00:00:00+00:00
99           119  ... 2019-03-06 00:00:00+00:00
100          120  ... 2018-08-06 00:00:00+00:00
101          121  ... 2017-09-13 00:00:00+00:00
102          123  ... 2018-12-30 00:00:00+00:00

[103 rows x 6 columns]

In [4]:
SELECT *  -- Select the records of all actors who play in a Comedy
FROM movies.actsin AS ai
LEFT JOIN movies.movies as m
on ai.movie_id=m.movie_id
WHERE m.genre = 'Comedy';

actsin_id  movie_id  actor_id  ...  runtime year_of_release renting_price
0           2        56         2  ...       91            2011          2.09
1           6        56         3  ...       91            2011          2.09
2           9         7         6  ...      110            2002          1.89
3          17        20        14  ...      135            2003          2.29
4          28         8        23  ...       91            2002          2.59
5          52        15        34  ...      128            2003          1.69
6          53        51        34  ...      107            2010          2.99
7          55        11        36  ...       95            2002          1.79
8          75        15        45  ...      128            2003          1.69
9          77         7        47  ...      110            2002          1.89
10         78        24        47  ...      110            2004          2.89
11         81         7        50  ...      110            2002          1.89
12         85        51        53  ...      107            2010          2.99
13         88         3        56  ...      127            2001          2.59
14         93        20        58  ...      135            2003          2.29
15        102        15        63  ...      128            2003          1.69
16        124        56        80  ...       91            2011          2.09
17        131        20        87  ...      135            2003          2.29
18        133         1        89  ...       93            2001          2.09
19        136        24        91  ...      110            2004          2.89
20        137         3        92  ...      127            2001          2.59
21        140         1        95  ...       93            2001          2.09
22        144         3        97  ...      127            2001          2.59
23        152         8       103  ...       91            2002          2.59
24        160         8       107  ...       91            2002          2.59
25        163         1       110  ...       93            2001          2.09
26        166        51       112  ...      107            2010          2.99
27        168        24       114  ...      110            2004          2.89
28        171        11       117  ...       95            2002          1.79
29        175        11       119  ...       95            2002          1.79

[30 rows x 9 columns]

In [5]:
SELECT * -- Create a list of all actors who play in a Comedy. Use the first letter of the table as an alias.
FROM movies.actors as a
WHERE EXISTS
	(SELECT *
	 FROM movies.actsin AS ai
	 LEFT JOIN movies.movies AS m
	 ON m.movie_id = ai.movie_id
	 WHERE m.genre = 'Comedy'
	 AND ai.actor_id = a.actor_id);

actor_id                name  year_of_birth       nationality  gender
0          2        Adam Sandler           1966               USA    male
1          3           Al Pacino           1940               USA    male
2          6     Anjelica Huston           1951               USA  female
3         14          Bill Nighy           1949           British    male
4         23     Charlize Theron           1975      South Africa  female
5         34        Diane Keaton           1946               USA  female
6         36        Eddie Murphy           1961               USA    male
7         45   Frances McDormand           1957               USA  female
8         47        Gene Hackman           1930               USA    male
9         50     Gwyneth Paltrow           1972               USA  female
10        53       Harrison Ford           1942               USA    male
11        56          Helen Hunt           1963               USA  female
12        58          Hugh Grant           1960           British    male
13        63      Jack Nicholson           1937               USA    male
14        80        Katie Holmes           1978               USA  female
15        87         Liam Neeson           1952  Northern Ireland    male
16        89           Liv Tyler           1977               USA  female
17        91   Marcia Gay Harden           1959               USA  female
18        92        Marisa Tomei           1964               USA  female
19        95         Matt Dillon           1964               USA    male
20        97          Mel Gibson           1956               USA    male
21       103  Natasha Richardson           1963           British  female
22       107      Patrick Swayze           1952               USA    male
23       110         Paul Reiser           1956               USA    male
24       112      Rachel McAdams           1978            Canada  female
25       114          Ray Romano           1957               USA    male
26       117          Rene Russo           1954               USA  female
27       119      Robert De Niro           1943               USA    male

In [6]:
SELECT a.nationality, count(*) -- Report the nationality and the number of actors for each nationality
FROM movies.actors AS a
WHERE EXISTS
	(SELECT ai.actor_id
	 FROM movies.actsin AS ai
	 LEFT JOIN movies.movies AS m
	 ON m.movie_id = ai.movie_id
	 WHERE m.genre = 'Comedy'
	 AND ai.actor_id = a.actor_id)
group by a.nationality;

nationality  count
0  Northern Ireland      1
1               USA     22
2      South Africa      1
3            Canada      1
4           British      3

## Queries with union and intersect  
Functions union and intersect.  
For the UNION it is crucial that the same columns are selected in both tables.  


In [1]:
SELECT name, 
       nationality, 
       year_of_birth
FROM movies.actors
WHERE nationality <> 'USA'
UNION-- Select all actors who are not from the USA and all actors who are born after 1990
SELECT name, 
       nationality, 
       year_of_birth
FROM movies.actors
WHERE year_of_birth > 1990;

name       nationality  year_of_birth
0       Cate Blanchett         Australia         1969.0
1       Julie Christie           British         1940.0
2          Emma Watson           British         1990.0
3            Sean Bean           British         1959.0
4     Freddie Highmore           British         1992.0
5       Eddie Redmayne           British         1982.0
6         Ian McKellen           British         1939.0
7          Colin Firth           British         1960.0
8        Tom Wilkinson           British         1948.0
9         Keanu Reeves            Canada         1964.0
10       Maryam Karimi                              NaN
11        Heath Ledger         Australia         1979.0
12     Christoph Waltz           Austria         1956.0
13          Bill Nighy           British         1949.0
14     Fernando Tielve             Spain         1986.0
15       Geoffrey Rush         Australia         1951.0
16      Rachel McAdams            Canada         1978.0
17      Jeremy Northam           British         1961.0
18      Liam Hemsworth         Australia         1990.0
19        Barkhad Abdi           Somalia         1985.0
20        Helen Mirren           British         1945.0
21        Hugh Jackman         Australia         1968.0
22   Donald Sutherland            Canada         1935.0
23       Javier Bardem             Spain         1969.0
24    Daniel Radcliffe           British         1989.0
25     Josh Hutcherson               USA         1992.0
26  Shohreh Aghdashloo              Iran         1952.0
27         Liam Neeson  Northern Ireland         1952.0
28  Natasha Richardson           British         1963.0
29       Nicole Kidman         Australia         1967.0
30   Carice van Houten       Netherlands         1976.0
31        Daniel Craig           British         1968.0
32        Ciaran Hinds  Northern Ireland         1953.0
33     Anthony Hopkins           British         1937.0
34        Barry Pepper            Canada         1970.0
35      Mohamad Dolati                              NaN
36      Christian Bale           British         1974.0
37       Michael Caine           British         1933.0
38       Jim Broadbent           British         1949.0
39    Benicio Del Toro       Puerto Rico         1967.0
40       Colin Farrell           Ireland         1976.0
41         Stephen Rea           Ireland         1946.0
42     Charlize Theron      South Africa         1975.0
43      Rachel Roberts           British         1927.0
44        Hugo Weaving         Australia         1960.0
45     Natalie Portman            Israel         1981.0
46         Naomi Watts           British         1968.0
47       Irene Escolar             Spain         1988.0
48       Abbie Cornish         Australia         1982.0
49     Annasophia Robb               USA         1993.0
50    Violante Placido             Italy         1976.0
51        Rupert Grint           British         1988.0
52      Hector Bordoni         Argentina         1965.0
53          Hugh Grant           British         1960.0
54  Andrea Riseborough           British         1981.0
55           Sam Riley           British         1980.0

In [2]:
SELECT name, 
       nationality, 
       year_of_birth
FROM movies.actors
WHERE nationality <> 'USA'
INTERSECT -- Select all actors who are not from the USA and who are also born after 1990
SELECT name, 
       nationality, 
       year_of_birth
FROM movies.actors
WHERE year_of_birth > 1990;

name nationality  year_of_birth
0  Freddie Highmore     British           1992

In [3]:
SELECT movie_id, avg(rating) -- Select the IDs of all movies with average rating higher than 9. Make a list of all movies that are in the drama genre and have an average rating higher than 9.
FROM movies.renting
GROUP BY movie_id
HAVING avg(rating) >9;

movie_id        avg
0        63   9.333333
1        42   9.600000
2        48  10.000000
3         5   9.750000

In [4]:
SELECT movie_id
FROM movies.movies
WHERE genre = 'Drama'
INTERSECT  -- Select the IDs of all dramas with average rating higher than 9
SELECT movie_id
FROM movies.renting
GROUP BY movie_id
HAVING AVG(rating)>9;

movie_id
0        42

In [5]:
SELECT *
FROM movies.movies
WHERE movie_id in -- Select all movies of genre drama with average rating higher than 9
   (SELECT movie_id
    FROM movies.movies
    WHERE genre = 'Drama'
    INTERSECT
    SELECT movie_id
    FROM movies.renting
    GROUP BY movie_id
    HAVING AVG(rating)>9);

movie_id                   title  ... year_of_release  renting_price
0        42  No Country for Old Men  ...            2007           1.49

[1 rows x 6 columns]

 ## OLAP: CUBE operator  
 Online analytical processing (or in shorth OLAP) as key technique to gain insight of the data.  
 The idea behind OLAP is to interactively analyze data and to summarize and visualize it.  
 Various OLAP operatros can be used to interactively analyze the data and look for interesting patterns.
 * Aggregate data for a better overview 
 * CUBE, ROLLUP, and GROUPING SETS operators
 * Produce pivot tables to present aggregation values

We can obtain aggregation values with the CUBE operator. It is always used in combination with the group by clause. For e.g.  
select country, genre, count(* )   
From renting:extended    
group by cube (country,genre)  

In [6]:
SELECT gender, -- Extract information of a pivot table of gender and country for the number of customers
	   country,
	   count(*)
FROM movies.customers
GROUP BY CUBE (gender, country)
ORDER BY country;

gender       country  count
0   female       Austria      3
1     male       Austria      1
2     None       Austria      4
3   female       Belgium      3
4     male       Belgium      3
5     None       Belgium      6
6   female       Denmark      4
7     male       Denmark      3
8     None       Denmark      7
9     male        France      8
10    None        France     13
11  female        France      5
12    male  Great Britan      5
13  female  Great Britan      5
14    None  Great Britan     10
15    male       Hungary      6
16  female       Hungary      5
17    None       Hungary     11
18    male         Italy     13
19    None         Italy     27
20  female         Italy     14
21    None        Poland     13
22  female        Poland      4
23    male        Poland      9
24  female      Slovenia      7
25    male      Slovenia      4
26    None      Slovenia     11
27  female         Spain     10
28    male         Spain      6
29    None         Spain     16
30    male           USA      2
31  female           USA      2
32    None           USA      4
33    male          None     60
34  female          None     62
35    None          None    122

In [7]:
-- List the number of movies for different genres and the year of release on all aggregation levels by using the CUBE operator.
SELECT genre,
       year_of_release,
       count(*)
FROM movies.movies
group by cube(genre,year_of_release)
ORDER BY year_of_release;

genre  year_of_release  count
0                        None           2001.0      6
1                       Drama           2001.0      2
2                      Comedy           2001.0      2
3   Science Fiction & Fantasy           2001.0      2
4                      Comedy           2002.0      3
..                        ...              ...    ...
57  Art House & International              NaN      1
58                      Drama              NaN     36
59                      Other              NaN      2
60                  Animation              NaN      2
61  Science Fiction & Fantasy              NaN     11

[62 rows x 3 columns]

In [8]:
-- Calculate the average rating for all aggregation levels of country and genre.
SELECT 
	c.country, 
	m.genre, 
	AVG(r.rating) AS avg_rating -- Calculate the average rating 
FROM movies.renting AS r
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
LEFT JOIN movies.customers AS c
ON r.customer_id = c.customer_id
group by cube(c.country,m.genre); -- For all aggregation levels of country and genre

country                      genre  avg_rating
0       None                       None    7.939024
1     France         Mystery & Suspense    6.000000
2   Slovenia         Action & Adventure         NaN
3      Spain                  Animation         NaN
4     Poland                     Comedy    8.200000
..       ...                        ...         ...
91      None  Art House & International    8.666667
92      None         Mystery & Suspense    6.833333
93      None                     Comedy    7.766667
94      None         Action & Adventure    9.090909
95      None  Science Fiction & Fantasy    8.163934

[96 rows x 3 columns]

The average over all records is 7.94 !

## ROLL UP  
The roll up operator is always used in combination with a group by statement. 
* It returns aggregates for a hierarchy of values.  
* In each step, one level of detail is dropped.  
* Then, order of column names is important for rollup.     

In [9]:
-- Count the total number of customers, the number of customers for each country, and the number of female and male customers for each country
SELECT country,
       gender,
	   COUNT(*)
FROM movies.customers
group by rollup (country, gender)
order by country, gender; -- Order the result by country and gender

country  gender  count
0        Austria  female      3
1        Austria    male      1
2        Austria    None      4
3        Belgium  female      3
4        Belgium    male      3
5        Belgium    None      6
6        Denmark  female      4
7        Denmark    male      3
8        Denmark    None      7
9         France  female      5
10        France    male      8
11        France    None     13
12  Great Britan  female      5
13  Great Britan    male      5
14  Great Britan    None     10
15       Hungary  female      5
16       Hungary    male      6
17       Hungary    None     11
18         Italy  female     14
19         Italy    male     13
20         Italy    None     27
21        Poland  female      4
22        Poland    male      9
23        Poland    None     13
24      Slovenia  female      7
25      Slovenia    male      4
26      Slovenia    None     11
27         Spain  female     10
28         Spain    male      6
29         Spain    None     16
30           USA  female      2
31           USA    male      2
32           USA    None      4
33          None    None    122

In [10]:
--Calculate the average ratings and the number of ratings for each country and each genre. Include the columns country and genre in the SELECT clause
SELECT 
	c.country, -- Select country
	m.genre, -- Select genre
	avg(r.rating), -- Average ratings
	count(*)  -- Count number of movie rentals
FROM movies.renting AS r
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
LEFT JOIN movies.customers AS c
ON r.customer_id = c.customer_id
GROUP BY country,genre -- Aggregate for each country and each genre
ORDER BY c.country, m.genre;

country                      genre        avg  count
0   Austria                     Comedy   8.000000      1
1   Austria                      Drama   6.000000      2
2   Austria         Mystery & Suspense        NaN      1
3   Austria  Science Fiction & Fantasy   6.666667      3
4   Belgium                     Comedy        NaN      1
..      ...                        ...        ...    ...
70      USA  Science Fiction & Fantasy   7.000000      2
71     None                  Animation  10.000000      1
72     None                     Comedy        NaN      1
73     None                      Drama   6.000000      2
74     None  Science Fiction & Fantasy        NaN      1

[75 rows x 4 columns]

In [11]:
-- Group by each county and genre with OLAP extension
--Finally, calculate the average ratings and the number of ratings for each country and genre, as well as an aggregation over all genres for each country and the overall average and total number.
SELECT 
	c.country, 
	m.genre, 
	AVG(r.rating) AS avg_rating, 
	COUNT(*) AS num_rating
FROM movies.renting AS r
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
LEFT JOIN movies.customers AS c
ON r.customer_id = c.customer_id
GROUP BY rollup(c.country,m.genre)
ORDER BY c.country, m.genre

country                      genre  avg_rating  num_rating
0   Austria                     Comedy    8.000000           1
1   Austria                      Drama    6.000000           2
2   Austria         Mystery & Suspense         NaN           1
3   Austria  Science Fiction & Fantasy    6.666667           3
4   Austria                       None    6.800000           7
..      ...                        ...         ...         ...
83     None                     Comedy         NaN           1
84     None                      Drama    6.000000           2
85     None  Science Fiction & Fantasy         NaN           1
86     None                       None    7.939024         578
87     None                       None    8.000000           5

[88 rows x 4 columns]

## OLAP operations: grouping sets

The most flexible OLAP operator is the GROUPING SETS operator.  
Example with a query with gropuing sets operator: 


select country, genre, count(* )  
from rentings_extended  
group by grouping sets((country,genre),(country),(genre),())  

* column names surrounded by parenthesis represent one level of aggregation.  
* GROUPING BY SETS returns an union over several group by queries.  
* It combines all information of a pivot table in one query.
* () means the total

The previous example represents the combination of 4 queries

In [12]:
-- Exploring nationality and gender of actors
-- Count the number of actors in the table actors from each country, the number of male and female actors and the total number of actors.
SELECT 
	nationality, -- Select nationality of the actors
    gender, -- Select gender of the actors
    count(*) -- Count the number of actors
FROM movies.actors
GROUP BY GROUPING SETS ((nationality), (gender), ()); -- Use the correct GROUPING SETS operation

nationality  gender  count
0               None    None    145
1          Argentina    None      1
2              Spain    None      3
3              Italy    None      1
4        Puerto Rico    None      1
5               Iran    None      1
6   Northern Ireland    None      2
7                USA    None     91
8             Israel    None      1
9        Netherlands    None      1
10      South Africa    None      1
11                      None      2
12           Austria    None      1
13         Australia    None      8
14           Ireland    None      2
15            Canada    None      4
16           British    None     24
17           Somalia    None      1
18              None    male     90
19              None  female     55

In [13]:
--  Use GROUPING SETS to get the average rating over country and gender.
SELECT 
	c.country, 
    c.gender,
	AVG(r.rating)
FROM movies.renting AS r
LEFT JOIN movies.customers AS c
ON r.customer_id = c.customer_id
group by GROUPING sets((country,gender)); -- Group by country and gender with GROUPING SETS

country  gender       avg
0        Austria    male  6.000000
1         France  female  8.000000
2        Hungary  female  7.285714
3           None    None  8.000000
4          Spain  female  7.612903
5        Belgium    male  8.000000
6            USA    male  7.500000
7        Denmark  female  8.444444
8        Austria  female  7.000000
9       Slovenia    male  8.090909
10       Belgium  female  9.125000
11      Slovenia  female  8.545455
12         Italy  female  8.366667
13       Denmark    male  7.333333
14  Great Britan    male  7.714286
15        France    male  7.666667
16        Poland  female  7.866667
17         Spain    male  7.681818
18         Italy    male  7.892857
19           USA  female  8.750000
20  Great Britan  female  7.272727
21       Hungary    male  7.857143
22        Poland    male  8.333333

In [14]:
-- Report all information that is included in a pivot table for country and gender in one SQL table.
SELECT 
	c.country, 
    c.gender,
	AVG(r.rating)
FROM movies.renting AS r
LEFT JOIN movies.customers AS c
ON r.customer_id = c.customer_id
-- Report all info from a Pivot table for country and gender
GROUP BY GROUPING SETS ((country, gender), (country),(gender),());

country  gender       avg
0           None    None  7.939024
1        Austria    male  6.000000
2         France  female  8.000000
3        Hungary  female  7.285714
4           None    None  8.000000
5          Spain  female  7.612903
6        Belgium    male  8.000000
7            USA    male  7.500000
8        Denmark  female  8.444444
9        Austria  female  7.000000
10      Slovenia    male  8.090909
11       Belgium  female  9.125000
12      Slovenia  female  8.545455
13         Italy  female  8.366667
14       Denmark    male  7.333333
15  Great Britan    male  7.714286
16        France    male  7.666667
17        Poland  female  7.866667
18         Spain    male  7.681818
19         Italy    male  7.892857
20           USA  female  8.750000
21  Great Britan  female  7.272727
22       Hungary    male  7.857143
23        Poland    male  8.333333
24          None    None  8.000000
25         Spain    None  7.641509
26  Great Britan    None  7.520000
27       Austria    None  6.800000
28        Poland    None  8.121212
29         Italy    None  8.137931
30      Slovenia    None  8.431818
31       Hungary    None  7.628571
32       Denmark    None  7.888889
33       Belgium    None  8.900000
34        France    None  7.771429
35           USA    None  8.000000
36          None    None  8.000000
37          None  female  8.052941
38          None    male  7.814103

The overall average of ratings is 7.9.

## Bringing all together  
* The first step is joining the data.  
* Select the relevant data.
* Choice the aggregation data and the level of aggregation.

Now the management considers investing money in movies of the best rated genres.

In [15]:
-- Augment the records of movie rentals with information about movies. Use the first letter of the table as alias.
SELECT *
FROM movies.renting AS r
left join movies.movies as m -- Augment the table with information about movies
on r.movie_id=m.movie_id;

renting_id  customer_id  movie_id  ...  runtime year_of_release  renting_price
0             1           41         8  ...       91            2002           2.59
1             2           10        29  ...      122            2005           2.79
2             3          108        45  ...       96            2008           2.39
3             4           39        66  ...      142            2012           1.59
4             5          104        15  ...      128            2003           1.69
..          ...          ...       ...  ...      ...             ...            ...
573         574            7        14  ...      115            2003           1.69
574         575           45        31  ...      157            2005           2.69
575         576           38        25  ...      108            2004           2.89
576         577           35        65  ...       96            2012           1.79
577         578           35        16  ...      135            2003           1.59

[578 rows x 11 columns]

In [16]:
--Select records of movies with at least 4 ratings, starting from 2018-04-01.
SELECT *
FROM movies.renting AS r
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
WHERE r.movie_id  IN ( -- Select records of movies with at least 4 ratings
	SELECT movie_id
	FROM movies.renting
	GROUP BY movie_id
	HAVING count(rating)>=4)
AND date_renting >= '2018-04-01'; -- Select records of movie rentals since 2018-04-01

renting_id  customer_id  movie_id  ...  runtime year_of_release  renting_price
0             3          108        45  ...       96            2008           2.39
1             4           39        66  ...      142            2012           1.59
2             6           50        71  ...      101            2014           1.79
3             8           73        65  ...       96            2012           1.79
4            12           52        65  ...       96            2012           1.79
..          ...          ...       ...  ...      ...             ...            ...
328         572            4        61  ...      130            2011           1.99
329         573          111        23  ...      106            2004           1.69
330         574            7        14  ...      115            2003           1.69
331         575           45        31  ...      157            2005           2.69
332         576           38        25  ...      108            2004           2.89

[333 rows x 11 columns]

In [17]:
-- For each genre, calculate the average rating (use the alias avg_rating), the number of ratings (use the alias n_rating), the number of movie rentals (use the alias n_rentals), and the number of distinct movies (use the alias n_movies).
SELECT m.genre, -- For each genre, calculate:
	   avg(r.rating) as avg_rating, -- The average rating and use the alias avg_rating
	   count(r.rating) as n_rating, -- The number of ratings and use the alias n_rating
	   count(*) as n_rentals,     -- The number of movie rentals and use the alias n_rentals
	   count(distinct m.movie_id) as n_movies -- The number of distinct movies and use the alias n_movies
FROM movies.renting AS r
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
WHERE r.movie_id IN ( 
	SELECT movie_id
	FROM movies.renting
	GROUP BY movie_id
	HAVING COUNT(rating) >= 3)
AND r.date_renting >= '2018-01-01'
GROUP BY GROUPING SETS ((m.genre));

genre  avg_rating  n_rating  n_rentals  n_movies
0         Action & Adventure    8.714286         7          9         2
1                  Animation    7.833333         6         10         2
2  Art House & International    8.500000         4          5         1
3                     Comedy    7.950000        20         31         5
4                      Drama    7.748252       143        245        34
5         Mystery & Suspense    7.428571         7         19         3
6                      Other    8.428571         7         16         2
7  Science Fiction & Fantasy    8.276596        47         70        10

In [18]:
-- Order the table by decreasing average rating.
SELECT genre,
	   AVG(rating) AS avg_rating,
	   COUNT(rating) AS n_rating,
       COUNT(*) AS n_rentals,     
	   COUNT(DISTINCT m.movie_id) AS n_movies 
FROM movies.renting AS r
LEFT JOIN movies.movies AS m
ON m.movie_id = r.movie_id
WHERE r.movie_id IN ( 
	SELECT movie_id
	FROM movies.renting
	GROUP BY movie_id
	HAVING COUNT(rating) >= 3 )
AND r.date_renting >= '2018-01-01'
GROUP BY genre
order by avg(rating) desc; -- Order the table by decreasing average rating


genre  avg_rating  n_rating  n_rentals  n_movies
0         Action & Adventure    8.714286         7          9         2
1  Art House & International    8.500000         4          5         1
2                      Other    8.428571         7         16         2
3  Science Fiction & Fantasy    8.276596        47         70        10
4                     Comedy    7.950000        20         31         5
5                  Animation    7.833333         6         10         2
6                      Drama    7.748252       143        245        34
7         Mystery & Suspense    7.428571         7         19         3

The last aspect you have to analyze are customer preferences for certain actors.

In [19]:
-- For each combination of the actors' nationality and gender, calculate the average rating, the number of ratings, the number of movie rentals, and the number of actors.
SELECT a.nationality,
       a.gender,
	   avg(r.rating) AS avg_rating, -- The average rating
	   count(r.rating) AS n_rating, -- The number of ratings
	   count(*) AS n_rentals, -- The number of movie rentals
	   count(distinct a.actor_id) AS n_actors -- The number of actors
FROM movies.renting AS r
LEFT JOIN movies.actsin AS ai
ON ai.movie_id = r.movie_id
LEFT JOIN movies.actors AS a
ON ai.actor_id = a.actor_id
WHERE r.movie_id IN ( 
	SELECT movie_id
	FROM movies.renting
	GROUP BY movie_id
	HAVING COUNT(rating) >=4 )
AND r.date_renting >= '2018-04-01'
GROUP BY GROUPING sets((a.nationality,a.gender)) ; -- Report results for each combination of the actors' nationality and gender

nationality  gender  avg_rating  n_rating  n_rentals  n_actors
0          Argentina    male    8.500000         4          5         1
1          Australia  female    8.666667         3          5         1
2          Australia    male    7.454545        11         17         3
3            Austria    male    8.500000         2          6         1
4            British  female    7.833333        54         78         3
5            British    male    8.105263       114        175         9
6             Canada  female    8.000000         5          5         1
7             Canada    male    8.428571         7         13         2
8            Ireland    male    7.714286         7         16         2
9             Israel  female    8.250000         4          8         1
10             Italy  female    9.000000         4          4         1
11  Northern Ireland    male    9.000000         4          4         1
12       Puerto Rico    male    7.200000         5          7         1
13           Somalia    male    7.000000         3          5         1
14      South Africa  female    8.214286        14         21         1
15             Spain  female    8.500000         4          5         1
16             Spain    male    8.857143         7         12         2
17               USA  female    7.545455       121        212        27
18               USA    male    7.913223       242        401        42

In [20]:
--  Provide results for all aggregation levels represented in a pivot table.  
SELECT a.nationality,
       a.gender,
	   AVG(r.rating) AS avg_rating,
	   COUNT(r.rating) AS n_rating,
	   COUNT(*) AS n_rentals,
	   COUNT(DISTINCT a.actor_id) AS n_actors
FROM movies.renting AS r
LEFT JOIN movies.actsin AS ai
ON ai.movie_id = r.movie_id
LEFT JOIN movies.actors AS a
ON ai.actor_id = a.actor_id
WHERE r.movie_id IN ( 
	SELECT movie_id
	FROM movies.renting
	GROUP BY movie_id
	HAVING COUNT(rating) >= 4)
AND r.date_renting >= '2018-04-01'
GROUP BY grouping sets ((a.nationality,a.gender),(a.nationality),(a.gender),()); -- Provide results for all aggregation levels represented in a pivot table


nationality  gender  avg_rating  n_rating  n_rentals  n_actors
0          Argentina    male    8.500000         4          5         1
1          Argentina    None    8.500000         4          5         1
2          Australia  female    8.666667         3          5         1
3          Australia    male    7.454545        11         17         3
4          Australia    None    7.714286        14         22         4
5            Austria    male    8.500000         2          6         1
6            Austria    None    8.500000         2          6         1
7            British  female    7.833333        54         78         3
8            British    male    8.105263       114        175         9
9            British    None    8.017857       168        253        12
10            Canada  female    8.000000         5          5         1
11            Canada    male    8.428571         7         13         2
12            Canada    None    8.250000        12         18         3
13           Ireland    male    7.714286         7         16         2
14           Ireland    None    7.714286         7         16         2
15            Israel  female    8.250000         4          8         1
16            Israel    None    8.250000         4          8         1
17             Italy  female    9.000000         4          4         1
18             Italy    None    9.000000         4          4         1
19  Northern Ireland    male    9.000000         4          4         1
20  Northern Ireland    None    9.000000         4          4         1
21       Puerto Rico    male    7.200000         5          7         1
22       Puerto Rico    None    7.200000         5          7         1
23           Somalia    male    7.000000         3          5         1
24           Somalia    None    7.000000         3          5         1
25      South Africa  female    8.214286        14         21         1
26      South Africa    None    8.214286        14         21         1
27             Spain  female    8.500000         4          5         1
28             Spain    male    8.857143         7         12         2
29             Spain    None    8.727273        11         17         3
30               USA  female    7.545455       121        212        27
31               USA    male    7.913223       242        401        42
32               USA    None    7.790634       363        613        69
33              None    None    7.902439       615        999       101
34              None  female    7.751196       209        338        36
35              None    male    7.980296       406        661        65